File: ui/dataset-explorer/backend/models/dataset_explorer.py
from typing import Annotated, Any, Literal

from pydantic import BaseModel, Field

from nl_code.datasets.task import CodeDataset


class DatasetOption(BaseModel):
    key: str
    dataset_id: CodeDataset
    label: str
    family: str
    split: str
    loaded: bool


class MetricDistribution(BaseModel):
    key: str
    label: str
    values: list[float]


class ScatterPoint(BaseModel):
    task_id: str
    x: float
    y: float


class MetricScatter(BaseModel):
    key: str
    label: str
    x_key: str
    x_label: str
    y_key: str
    y_label: str
    points: list[ScatterPoint]


class FlawedErrorGroup(BaseModel):
    error: str
    count: int
    task_ids: list[str]


class OverviewCounts(BaseModel):
    raw_sample_count: int
    task_count: int
    flawed_count: int


class DatasetOverviewResponse(BaseModel):
    dataset: DatasetOption
    counts: OverviewCounts
    distributions: list[MetricDistribution]
    scatter_plots: list[MetricScatter]
    flawed_error_groups: list[FlawedErrorGroup]


class DatasetRefreshResponse(BaseModel):
    dataset: DatasetOption
    reloaded: bool


class SummaryStats(BaseModel):
    count: int
    min: float
    median: float
    p90: float
    max: float


class MetricSummary(BaseModel):
    key: str
    label: str
    stats: SummaryStats


class RatioSummary(BaseModel):
    key: str
    label: str
    numerator_key: str
    denominator_key: str
    stats: SummaryStats


class CrossDatasetSeriesValues(BaseModel):
    dataset_key: str
    dataset_label: str
    values: list[float]


class CrossDatasetSeries(BaseModel):
    key: str
    label: str
    datasets: list[CrossDatasetSeriesValues]


class DatasetLandscapePoint(BaseModel):
    dataset_key: str
    dataset_label: str
    family: str
    task_count: int
    median_prompt_length_chars: float
    median_derived_code_length_chars: float


class DatasetCompareRow(BaseModel):
    dataset: DatasetOption
    counts: OverviewCounts
    flawed_rate: float
    metrics: list[MetricSummary]
    ratios: list[RatioSummary]


class DatasetCompareResponse(BaseModel):
    datasets: list[DatasetCompareRow]
    metric_series: list[CrossDatasetSeries]
    ratio_series: list[CrossDatasetSeries]
    landscape_points: list[DatasetLandscapePoint]


class TaskRow(BaseModel):
    task_id: str
    status: Literal["valid", "flawed"]
    has_derived_task: bool
    entry_point_name: str | None = None
    description_preview: str | None = None
    description_length_chars: int | None = None
    description_length_tokens: int | None = None
    derived_code_length_chars: int | None = None
    derived_code_length_tokens: int | None = None
    derived_code_length_lines: int | None = None
    prompt_length_chars: int | None = None
    raw_source_length_chars: int | None = None
    test_length_chars: int | None = None
    error_summary: str | None = None


class TaskListResponse(BaseModel):
    dataset: DatasetOption
    total: int
    page: int
    per_page: int
    rows: list[TaskRow]


class NumericMetric(BaseModel):
    key: str
    label: str
    value: float


class DerivedFieldSummary(BaseModel):
    name: str
    value: str
    source: str


class TaskDetailResponse(BaseModel):
    dataset: DatasetOption
    task_id: str
    entry_point_name: str
    description: str
    gt_solution: str
    metrics: list[NumericMetric]
    derived_fields: list[DerivedFieldSummary]
    prev_task_id: str | None
    next_task_id: str | None


class InspectorField(BaseModel):
    key: str
    label: str
    kind: Literal["text", "code", "json", "list", "error"]
    value: Any


class InspectorSection(BaseModel):
    title: str
    fields: list[InspectorField]


class RawDetailBase(BaseModel):
    dataset: DatasetOption
    task_id: str
    title: str
    validated: bool | None
    sections: list[InspectorSection]
    derived_fields: list[DerivedFieldSummary]
    raw_json: dict[str, Any]
    prev_task_id: str | None
    next_task_id: str | None


class HumanEvalRawDetail(RawDetailBase):
    detail_kind: Literal["humaneval_raw_detail"]


class ProRawDetail(RawDetailBase):
    detail_kind: Literal["pro_raw_detail"]


class ClassEvalRawDetail(RawDetailBase):
    detail_kind: Literal["classeval_raw_detail"]


class FlawedRawDetail(BaseModel):
    detail_kind: Literal["flawed_raw_detail"]
    dataset: DatasetOption
    task_id: str
    title: str
    error: str
    sections: list[InspectorSection]
    raw_json: dict[str, Any]
    prev_task_id: str | None
    next_task_id: str | None


RawDetailResponse = Annotated[
    HumanEvalRawDetail | ProRawDetail | ClassEvalRawDetail | FlawedRawDetail,
    Field(discriminator="detail_kind"),
]

File: ui/dataset-explorer/frontend/src/types/datasetExplorer.ts
export interface DatasetOption {
  key: string;
  dataset_id: string;
  label: string;
  family: string;
  split: string;
  loaded: boolean;
}

export interface MetricDistribution {
  key: string;
  label: string;
  values: number[];
}

export interface ScatterPoint {
  task_id: string;
  x: number;
  y: number;
}

export interface MetricScatter {
  key: string;
  label: string;
  x_key: string;
  x_label: string;
  y_key: string;
  y_label: string;
  points: ScatterPoint[];
}

export interface FlawedErrorGroup {
  error: string;
  count: number;
  task_ids: string[];
}

export interface OverviewCounts {
  raw_sample_count: number;
  task_count: number;
  flawed_count: number;
}

export interface DatasetOverviewResponse {
  dataset: DatasetOption;
  counts: OverviewCounts;
  distributions: MetricDistribution[];
  scatter_plots: MetricScatter[];
  flawed_error_groups: FlawedErrorGroup[];
}

export interface SummaryStats {
  count: number;
  min: number;
  median: number;
  p90: number;
  max: number;
}

export interface MetricSummary {
  key: string;
  label: string;
  stats: SummaryStats;
}

export interface RatioSummary {
  key: string;
  label: string;
  numerator_key: string;
  denominator_key: string;
  stats: SummaryStats;
}

export interface CrossDatasetSeriesValues {
  dataset_key: string;
  dataset_label: string;
  values: number[];
}

export interface CrossDatasetSeries {
  key: string;
  label: string;
  datasets: CrossDatasetSeriesValues[];
}

export interface DatasetLandscapePoint {
  dataset_key: string;
  dataset_label: string;
  family: string;
  task_count: number;
  median_prompt_length_chars: number;
  median_derived_code_length_chars: number;
}

export interface DatasetCompareRow {
  dataset: DatasetOption;
  counts: OverviewCounts;
  flawed_rate: number;
  metrics: MetricSummary[];
  ratios: RatioSummary[];
}

export interface DatasetCompareResponse {
  datasets: DatasetCompareRow[];
  metric_series: CrossDatasetSeries[];
  ratio_series: CrossDatasetSeries[];
  landscape_points: DatasetLandscapePoint[];
}

export interface TaskRow {
  task_id: string;
  status: "valid" | "flawed";
  has_derived_task: boolean;
  entry_point_name: string | null;
  description_preview: string | null;
  description_length_chars: number | null;
  description_length_tokens: number | null;
  derived_code_length_chars: number | null;
  derived_code_length_tokens: number | null;
  derived_code_length_lines: number | null;
  prompt_length_chars: number | null;
  raw_source_length_chars: number | null;
  test_length_chars: number | null;
  error_summary: string | null;
}

export interface TaskListResponse {
  dataset: DatasetOption;
  total: number;
  page: number;
  per_page: number;
  rows: TaskRow[];
}

export interface NumericMetric {
  key: string;
  label: string;
  value: number;
}

export interface DerivedFieldSummary {
  name: string;
  value: string;
  source: string;
}

export interface TaskDetailResponse {
  dataset: DatasetOption;
  task_id: string;
  entry_point_name: string;
  description: string;
  gt_solution: string;
  metrics: NumericMetric[];
  derived_fields: DerivedFieldSummary[];
  prev_task_id: string | null;
  next_task_id: string | null;
}

export interface InspectorField {
  key: string;
  label: string;
  kind: "text" | "code" | "json" | "list" | "error";
  value: unknown;
}

export interface InspectorSection {
  title: string;
  fields: InspectorField[];
}

export interface RawDetailBase {
  dataset: DatasetOption;
  task_id: string;
  title: string;
  validated: boolean | null;
  sections: InspectorSection[];
  derived_fields: DerivedFieldSummary[];
  raw_json: Record<string, unknown>;
  prev_task_id: string | null;
  next_task_id: string | null;
}

export interface HumanEvalRawDetail extends RawDetailBase {
  detail_kind: "humaneval_raw_detail";
}

export interface ProRawDetail extends RawDetailBase {
  detail_kind: "pro_raw_detail";
}

export interface ClassEvalRawDetail extends RawDetailBase {
  detail_kind: "classeval_raw_detail";
}

export interface FlawedRawDetail {
  detail_kind: "flawed_raw_detail";
  dataset: DatasetOption;
  task_id: string;
  title: string;
  error: string;
  sections: InspectorSection[];
  raw_json: Record<string, unknown>;
  prev_task_id: string | null;
  next_task_id: string | null;
}

export type RawDetailResponse =
  | HumanEvalRawDetail
  | ProRawDetail
  | ClassEvalRawDetail
  | FlawedRawDetail;
